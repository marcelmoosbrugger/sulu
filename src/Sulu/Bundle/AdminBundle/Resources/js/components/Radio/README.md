Radio buttons keep no internal state and have to be managed from the outside, like shown in the
following example:

```
initialState = {value: '1'};
<div>
    <Radio checked={state.value === '1'} onChange={() => setState({value: '1'})} />
    <Radio checked={state.value === '2'} onChange={() => setState({value: '2'})} />
    <Radio checked={state.value === '3'} onChange={() => setState({value: '3'})} />
</div>
```

In most cases the state management of the radio buttons will be the same.
For that matter the `RadioGroup` component makes the use of the radio buttons more convenient.
```
const RadioGroup = require('./RadioGroup').default;

initialState = {value: '1'};

<RadioGroup value={state.value} onChange={(value) => setState({value})}>
    <Radio value="1" />
    <Radio value="2" />
    <Radio value="3" />
</RadioGroup>
```
