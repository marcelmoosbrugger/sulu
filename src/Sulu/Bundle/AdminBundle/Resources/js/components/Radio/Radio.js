// @flow
import React from 'react';
import type {ElementRef, Node} from 'react';
import radioStyles from './radio.scss';

type Props = {
    checked: boolean,
    value: string,
    onChange?: (value: string) => void,
    children: Node,
};

export default class Radio extends React.PureComponent<Props> {
    static defaultProps = {
        value: 'on',
        checked: false,
    };

    input: ElementRef<'input'>;

    handleChange = () => {
        if (this.props.onChange) {
            this.props.onChange(this.props.value);
        }
    };

    handleClick = () => this.input.click();
    handleInputClick = (event: Event) => event.stopPropagation();
    setInput = (input: ElementRef<'input'>) => this.input = input;

    render() {
        return (
            <label className={radioStyles.radio}>
                <span onClick={this.handleClick}>
                    <input
                        ref={this.setInput}
                        type="radio"
                        checked={this.props.checked}
                        onClick={this.handleInputClick}
                        onChange={this.handleChange} />
                    <span />
                </span>
                {this.renderChildren()}
            </label>
        );
    }

    renderChildren() {
        if (this.props.children) {
            return <span>{this.props.children}</span>;
        }

        return null;
    }
}
