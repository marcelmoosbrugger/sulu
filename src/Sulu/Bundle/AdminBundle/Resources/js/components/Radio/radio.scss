@import '../../containers/Application/colors.scss';

$radius: 8px;
$color: $blueZodiac;
$borderWidth: 3px;
$borderColor: $white;
$marginBottom: 10px;
$fontSize: 12px;

.radio {
    display: flex;
    margin-bottom: $marginBottom;
    align-items: center;
    cursor: pointer;

    & > span:nth-child(1) {
        display: block;
        width: calc(2 * $radius);
        height: calc(2 * $radius);
        position: relative;
    }

    & > span:nth-child(2) {
        display: block;
        margin-left: 10px;
        font-size: $fontSize;
    }

    input {
        position: relative;
        width: 100%;
        height: 100%;
        opacity: 0;
        cursor: pointer;
        z-index: 1;
    }

    input + span {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 100%;
        border-radius: $radius;
        background: $wildSand;
        border: $borderWidth solid $borderColor;
    }

    input:checked + span {
        background: $color;
    }
}
