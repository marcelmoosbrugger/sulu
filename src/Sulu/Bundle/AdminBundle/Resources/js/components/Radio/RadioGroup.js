// @flow
import React from 'react';
import type {ChildrenArray} from 'react';
import Radio from './Radio';

type Props = {
    value: string,
    onChange?: () => void,
    className?: string,
    children?: ChildrenArray<*>,
};

export default class RadioGroup extends React.PureComponent<Props> {
    addPropsToRadioChildren(children: ChildrenArray<*>) {
        return React.Children.map(children, (child) => {
            if (child.type === Radio) {
                return React.cloneElement(child, {
                    checked: this.props.value && child.props.value === this.props.value,
                    onChange: this.props.onChange,
                });
            }
            return React.cloneElement(child, {
                children: child.props.children ? this.addPropsToRadioChildren(child.props.children) : null,
            });
        });
    }

    render() {
        return (
            <div className={this.props.className}>
                {this.addPropsToRadioChildren(this.props.children)}
            </div>
        );
    }
}
