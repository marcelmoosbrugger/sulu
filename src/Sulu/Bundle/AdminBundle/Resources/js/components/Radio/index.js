// @flow
import Radio from './Radio';
import RadioGroup from './RadioGroup';

export {RadioGroup, Radio};
