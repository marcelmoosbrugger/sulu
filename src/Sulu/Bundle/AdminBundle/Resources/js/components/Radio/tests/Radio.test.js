/* eslint-disable flowtype/require-valid-file-annotation */
import {shallow, render} from 'enzyme';
import React from 'react';
import Radio from '../Radio';

test('The component should render in unchecked state', () => {
    const radio = render(<Radio checked={false}>My radio</Radio>);
    expect(radio).toMatchSnapshot();
});

test('The component should render in checked state', () => {
    const radio = render(<Radio checked={true}>My radio</Radio>);
    expect(radio).toMatchSnapshot();
});

test('The component should render without children', () => {
    const radio = render(<Radio checked={true} />);
    expect(radio).toMatchSnapshot();
});

test('A click on the button should trigger the change callback', () => {
    const onChangeSpy = jest.fn();
    const radio = shallow(<Radio value="my-radio" checked={false} onChange={onChangeSpy}>My radio</Radio>);
    radio.find('input').simulate('change');
    expect(onChangeSpy).toHaveBeenCalledWith('my-radio');
});
