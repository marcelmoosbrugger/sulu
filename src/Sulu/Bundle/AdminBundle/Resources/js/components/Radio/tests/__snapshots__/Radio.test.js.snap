// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should render in checked state 1`] = `
<label
  class="radio"
>
  <span>
    <input
      checked=""
      type="radio"
    />
    <span />
  </span>
  <span>
    My radio
  </span>
</label>
`;

exports[`The component should render in unchecked state 1`] = `
<label
  class="radio"
>
  <span>
    <input
      type="radio"
    />
    <span />
  </span>
  <span>
    My radio
  </span>
</label>
`;

exports[`The component should render without children 1`] = `
<label
  class="radio"
>
  <span>
    <input
      checked=""
      type="radio"
    />
    <span />
  </span>
</label>
`;
